### PR TITLE
Reader refactor and SPDX 2.3 reader

### PR DESCRIPTION
**Pull request by puerco**
After the cohort meeting last week and some discussions about the library architecture with @houdini91 this PR refactors the reader package to use the same pattern as the writer: We now have a `reader.Reader` object that reads SBOMs by using objects that implement the `Unserializer` interface.

The reader architecture is pictured more or less this ugly mermaid diagram

```mermaid
flowchart TD
    C[Reader] --> protobom
    File["fa:fa-file file.[cdx|spdx].json"] --> OpenFile 
    SPDX23Unserializer --"ParseStream()"---> C
    CycloneDX14Unserializer --"ParseStream()"---> C
    subgraph r [reader]
    OpenFile --> C 
    C <--> S["SniffFormat()"]    
end
```

This PR also breaks the dependency on my onesbom library which I used to build the initial POC. To do that, protobom now has its own format sniffer and a brand new SPDX reader based on the official [SPDX tools-golang](https://github.com/spdx/tools-golang) libraries.

The line count in the PR is quite high because there is an update to go.mod/go.sum and I'm checking in samples and test fixtures. Nevertheless, I tried to make the commits atomic enough to be easy to review one by one.

/cc @houdini91  (for some reason I can't assign you on th PR)

### CODE REVIEW

**Comment by jspeed-meyers** [comment]
cc @houdini91

File: pkg/formats/sniffer.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+func (fs *Sniffer) SniffReader(f io.ReadSeeker) (Format, error) {
+	defer f.Seek(0, 0)
+	fileScanner := bufio.NewScanner(f)
+	fileScanner.Split(bufio.ScanLines)
+
+	formatType := ""
+	formatEncoding := ""
+	formatVersion := ""
+
+	for fileScanner.Scan() {
```
Comment (jspeed-meyers) [inline pkg/formats/sniffer.go:32]: @bardenstein, I think you are interested in this type of format autodetection. It's code, so everything can be changed :) But if you wanted to weigh on any other heuristics you might want, feel free to. You can also open issues if you would prefer others.

File: pkg/formats/sniffer_test.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
+package formats
```
Comment (jspeed-meyers) [inline pkg/formats/sniffer_test.go:1]: I love tests! Nice.

File: pkg/reader/reader_spdx23.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+	// TODO(degradation): External document references
+
+	// TODO(puerco) Top level elements
+	if t := fp.spdxDateToTime(spdxDoc.CreationInfo.Created); t != nil {
+		bom.Metadata.Date = timestamppb.New(*t)
+	}
+	if spdxDoc.CreationInfo.Creators != nil {
+		for _, c := range spdxDoc.CreationInfo.Creators {
+			// TODO: We need to creaste a parser library in formats/dpsx
```
Comment (jspeed-meyers) [inline pkg/reader/reader_spdx23.go:39]:
```suggestion
			// TODO: We need to create a parser library in formats/dpsx
```

File: pkg/reader/reader_spdx23.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+			bom.NodeList.RootElements = append(bom.NodeList.RootElements, string(r.RefB.ElementRefID))
+		} else {
+			bom.NodeList.AddEdge(fp.relationshipToEdge(r))
+		}
+	}
+
+	return bom, nil
+}
+
+// packageToNode assigns the data from an SPDX package into a
```
Comment (jspeed-meyers) [inline pkg/reader/reader_spdx23.go:73]: Into a what?

**Review by jspeed-meyers — APPROVED** [review]
Silly nit comments. Nice work.

**Review by houdini91 — APPROVED** [review]
LGTM, added some general comments,
I also pushed a PR to the feature branch on your fork with a cleanup suggestion to the sniffer logic.

**Comment by houdini91** [comment]
@puerco @jspeed-meyers 
Can we call the reader - parser, `unserilizer` instead?
In my head it the mirror definition of writer to serilizer better.

Also @puerco i think we should merge the cdx serlizer cleanup PR and this PR before we add other parser/unserlizers missing.

Lastly should we move the `Translate` func in to a high level API in the packagem or for now we can leave it external to the lib?

**Comment by houdini91** [comment]
@jspeed-meyers https://github.com/puerco/protobom/pull/2

File: cmd/cli/root.go [hunk 5]
```diff
@@ -33,10 +33,9 @@ var (
 )
 
 var RootCmd = &cobra.Command{
-	Version: version,
```
[the remark this answers is not in the document]
Comment (puerco) [reply cmd/cli/root.go:36]: It's conflicting with the version flag you added below and it doesn't run :)

File: pkg/reader/reader_spdx23.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+
+	// TODO(puerco) Top level elements
+	if t := fp.spdxDateToTime(spdxDoc.CreationInfo.Created); t != nil {
+		bom.Metadata.Date = timestamppb.New(*t)
+	}
+	if spdxDoc.CreationInfo.Creators != nil {
+		for _, c := range spdxDoc.CreationInfo.Creators {
+			// TODO: We need to creaste a parser library in formats/dpsx
+			if c.CreatorType == "Tool" {
+				// TODO:
```
[the remark this answers is not in the document]
Comment (puerco) [reply pkg/reader/reader_spdx23.go:41]: Added

**Comment by puerco** [comment]
@houdini91 sounds good! I have renamed Parser Unserializer and updated things accordingly.

As I commented on [puerco/protobom#2](https://github.com/puerco/protobom/pull/2) let's tackle the sniffer logic patch in a new PR against main and we can discuss it there (check the comments I left). I'll review the cdx PR.

Let's get the translation package going it is really the last layer before getting to the cli :rocket:

**Pull request merged**